crates/holochain_persistence_lmdb/src/common.rs

```diff
@@ -1,48 +1,111 @@
-use holochain_logging::prelude::*;
-use lmdb::Error as LmdbError;
+//use holochain_logging::prelude::*;
+//use lmdb::Error as LmdbError;
+use crate::error::{is_store_full_error, is_store_full_result};
+use holochain_persistence_api::{
+    cas::content::{Address, Content},
+    hash::HashString,
+};
 use rkv::{
-    DatabaseFlags, EnvironmentFlags, Manager, Rkv, SingleStore, StoreError, StoreOptions, Value,
+    DatabaseFlags, EnvironmentFlags, Manager, Reader, Rkv, SingleStore, StoreError, StoreOptions,
+    Value, Writer,
 };
 use std::{
+    collections::HashMap,
     path::Path,
     sync::{Arc, RwLock},
 };
 
-const DEFAULT_INITIAL_MAP_BYTES: usize = 100 * 1024 * 1024;
+const DEFAULT_INITIAL_MAP_SIZE: usize = 100 * 1024 * 1024;
 
 #[derive(Clone)]
-pub(crate) struct LmdbInstance {
-    pub store: SingleStore,
-    pub manager: Arc<RwLock<Rkv>>,
+pub struct LmdbInstance {
+    store: SingleStore,
+    /// A lock protected reference to the environment. This lock
+    /// is a consequence of using the `rkv::Manager` api which returns
+    /// guarded environments to protect users from creating multiple environments
+    /// in the same process and path.
+    rkv: Arc<RwLock<Rkv>>,
 }
 
 impl LmdbInstance {
     pub fn new<P: AsRef<Path> + Clone>(
         db_name: &str,
         path: P,
         initial_map_bytes: Option<usize>,
+        flags: Option<EnvironmentFlags>,
     ) -> LmdbInstance {
-        let db_path = path.as_ref().join(db_name).with_extension("db");
-        std::fs::create_dir_all(db_path.clone()).expect("Could not create file path for store");
+        Self::new_all(&[db_name], path, initial_map_bytes, flags, None)
+            .into_iter()
+            .next()
+            .map(|(_db_name, instance)| instance)
+            .expect("Expected exactly one database instance")
+    }
+
+    /// Instantiates multiple lmdb database instances for a set of `db_names` at `path` and
+    /// initial map size `initial_map_size`. Transactions will be synchronized over all of them.
+    ///
+    /// If `use_rkv_manager` is set to the default of `true`, the provided `rkv::Manager` singleton
+    /// will ensure only one environment is created for the process per given `path`. If this is
+    /// already enforced outside this api then set it to `false`.
+    pub fn new_all<P: AsRef<Path> + Clone>(
+        db_names: &[&str],
+        path: P,
+        initial_map_size: Option<usize>,
+        flags: Option<EnvironmentFlags>,
+        use_rkv_manager: Option<bool>,
+    ) -> HashMap<String, LmdbInstance> {
+        std::fs::create_dir_all(path.clone()).expect("Could not create file path for store");
 
-        let manager = Manager::singleton()
-            .write()
-            .unwrap()
-            .get_or_create(db_path.as_path(), |path: &Path| {
+        let rkv =
+            if use_rkv_manager.unwrap_or_else(|| true) {
```

Comment:
Lots of duplicated code here: should make a private function to call within the if/else block.